include/odp/api/spec/classification.h

```diff
@@ -605,6 +619,28 @@ typedef struct odp_pmr_param_t {
 
 } odp_pmr_param_t;
 
+/**
+ * Packet Matching Rule creation options
+ */
+typedef struct odp_pmr_create_opt_t {
+	/** Array of odp_pmr_param_t entries, one entry per term desired.
+	 * Use odp_cls_pmr_param_init() to initialize parameters into their default values.
+	 */
+	odp_pmr_param_t *terms;
+	/** Number of terms in the match rule. */
+	int num_terms;
+	/** When a PMR is created with mark support, the implementation shall set
+	 * the specified mark in the packet metadata if PMR matches. In the event
+	 * of not matching the PMR, implementation set the ODP_COS_MARK_INVALID in
+	 * the packet metadata. The application can retrieve mark value set by
+	 * the implementation through odp_packet_mark().
+	 * The default value is ODP_COS_MARK_INVALID which means marking is
+	 * disabled. Application can set any value from ODP_COS_MARK_INVALID(0) to
+	 * odp_cls_capability_t::max_mark
+	 */
+	uint64_t mark;
```

Comment:
This also get shorter if invalid and enable/disable specification is removed.

- default / initial value is zero
- this value replaces current mark value when rule matches
- user can use values 0 ... max
- what happens when rule does not match? Current value is overwritten with 0, or not changed

Reply:
If a rule does not match, then I think the mark that is already in the packet should not change in any way .

The open issue that we have about the semantics and that Petri refers to above an in other comments is whether we have
1) Only one kind of PMR rules, that always set the mark upon rule match. The mark value of the last matching rule is the one that the application will see. Rules created with the old create command are equivalent to rules created with the create_opt() command with mark value zero.
2) Two kind of PMT rules: rules that set the mark and rules that do not. Rules created with the old command or with mark value zero do not change the mark of the packet. The mark that the application sees is the mark value of that last matching rule that had marking enabled, or the default mark value if no such rule matched. Rules created with the old command are equivalent to rules created with the new command with mark value 0.

The two alternatives are equivalent if there is only one layer of rules, but not if there is a multi level hierarchy of rules.

Which one do we want? If we want the latter, then the mark-setting-enable config could be a separate flag in the opt struct instead of having a special mark value designate it. And if so, then alternative 2 can be seen as a superset of alternative 1 and can be added later in a backward compatible manner.

Reply:
We can stick with option 1. I will send the version based on that.

Reply:
Updated to following in the next version.
        /** The mark value to retrieve it through odp_packet_cls_mark() on PMR match.
         * On PMR miss, odp_packet_cls_mark() shall return ODP_COS_MARK_DEFAULT.
         * The default mark value is ODP_COS_MARK_DEFAULT.
         * 0(ODP_COS_MARK_DEFAULT) to odp_cls_capability_t::max_mark is the valid
         * value range for mark.
         */